src/winconpty/winconpty.cpp

```diff
@@ -0,0 +1,383 @@
+    pi.hProcess = NULL;
+
+    RETURN_IF_NTSTATUS_FAILED(CreateClientHandle(&pPty->hPtyReference,
+                                                 serverHandle.get(),
+                                                 L"\\Reference",
+                                                 FALSE));
+
+    pPty->hSignal = signalPipeOurSide.release();
+
+    return S_OK;
+}
+
+// Function Description:
+// - Resizes the conpty
+// Arguments:
+// - hSignal: A signal pipe as returned by CreateConPty.
+// - size: The new dimenstions of the conpty, in characters.
+// Return Value:
+// - S_OK if the call succeeded, else an appropriate HRESULT for failing to
+//      write the resize message to the pty.
+HRESULT _ResizePseudoConsole(_In_ const PseudoConsole* const pPty, _In_ const COORD size)
+{
+    if (pPty == NULL)
+    {
+        return E_INVALIDARG;
+    }
+
+    unsigned short signalPacket[3];
+    signalPacket[0] = PTY_SIGNAL_RESIZE_WINDOW;
+    signalPacket[1] = size.X;
+    signalPacket[2] = size.Y;
+
+    BOOL fSuccess = WriteFile(pPty->hSignal, signalPacket, sizeof(signalPacket), NULL, NULL);
+    return fSuccess ? S_OK : HRESULT_FROM_WIN32(GetLastError());
+}
+
+// Function Description:
+// - This closes each of the members of a PseudoConsole. It does not free the
+//      data associated with the PseudoConsole. This is helpful for testing,
+//      where we might stack allocate a PseudoConsole (instead of getting a
+//      HPCON via the API).
+// Arguments:
+// - pPty: A pointer to a PseudoConsole struct.
+// Return Value:
+// - <none>
+void _ClosePseudoConsoleMembers(_In_ PseudoConsole* pPty)
+{
+    if (pPty != NULL)
+    {
+        // See MSFT:19918626
+        // First break the signal pipe - this will trigger conhost to tear itself down
+        if (_HandleIsValid(pPty->hSignal))
+        {
+            CloseHandle(pPty->hSignal);
+            pPty->hSignal = 0;
+        }
+        // Then, wait on the conhost process before killing it.
+        // We do this to make sure the conhost finishes flushing any output it
+        //      has yet to send before we hard kill it.
+        if (_HandleIsValid(pPty->hConPtyProcess))
+        {
+            // If the conhost is already dead, then that's fine. Presumably
+            //      it's finished flushing it's output already.
+            DWORD dwExit = 0;
+            // If GetExitCodeProcess failed, it's likely conhost is already dead
+            //      If so, skip waiting regardless of whatever error
+            //      GetExitCodeProcess returned.
+            //      We'll just go straight to killing conhost.
+            if (GetExitCodeProcess(pPty->hConPtyProcess, &dwExit) && dwExit == STILL_ACTIVE)
+            {
+                WaitForSingleObject(pPty->hConPtyProcess, INFINITE);
+            }
+
+            TerminateProcess(pPty->hConPtyProcess, 0);
+            pPty->hConPtyProcess = 0;
+        }
+        // Then take care of the reference handle.

```

Comment:
Can you add/leave the todo here that holding onto the reference handle until the end makes conhost think it has an outstanding client and I had to do a lot of dance to resolve that (and we should maybe reorder this in the future?)